### PR TITLE
[FLINK-11773][core] Harden KryoSerializerSnapshot

### PR DESCRIPTION
**Pull request by igalshilman**
This PR builds on top of #7845 .

## What is the purpose of the change

`KryoSerializerSnapshot` would fail in `readSnapshot` if a previously registered `Kryo` type is not accessible by the restored job anymore (either removed from the classpath, or failed to load for some other reason)

The source of this behaviour is an early return at [1], which would skip the consumption at [2] 
 
 [1] https://github.com/apache/flink/blob/cf7b86de436c8714414f563e8637ceb36ea7aabe/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotData.java#L310

[2] https://github.com/apache/flink/blob/cf7b86de436c8714414f563e8637ceb36ea7aabe/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotData.java#L314

## Brief change log
* The first two commits are from  #7845 and they are used by 3224e60 (KryoSnapshotTest)
* 2b95d56 adds an `LinkedOptionalMapSerializer` to robustly read/write `LinkedOptionalMaps`
* b4b47ef uses `LinkedOptionalMapSerializer` by `KryoSnapshotData` and `PojoSnapshotData`

## Verifying this change

* A new unit test for `KryoSerializerSnapshot` was introduced.

This change added tests and can be verified as follows:

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by igalshilman** [comment]
@flinkbot attention @tzulitai

**Comment by tzulitai** [comment]
@flinkbot approve description
@flinkbot approve architecture
@flinkbot approve consensus

File: flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * LinkedOptionalMapSerializer - A serializer of {@link LinkedOptionalMap}.
+ */
+@Internal
+public final class LinkedOptionalMapSerializer {
+
+	private static final long HEADER = 0x4f2c69a3d70L;
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java]: nit: would add a comment about what the HEADER is for. I guess it is for sanity checks?

File: flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			else {
+				value = null;
+			}
+
+			map.put(keyName, key, value);
+		}
+		return map;
+	}
+
+	private static <T> void writeFramed(DataOutputView out, BiConsumerWithException<DataOutputView, T, IOException> writter, T item) throws IOException {
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java]: nit: typo, `writter` -> `writer`

File: flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		ByteArrayOutputStreamWithPos streamWithPos = new ByteArrayOutputStreamWithPos();
+		try (DataOutputViewStreamWrapper inMemoryStream = new DataOutputViewStreamWrapper(streamWithPos)) {
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java]: I think you can actually use `DataOutputSerializer` here.

File: flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		out.writeInt(streamWithPos.getPosition());
+		out.write(streamWithPos.getBuf(), 0, streamWithPos.getPosition());
+	}
+
+	@Nullable
+	private static <T> T tryReadFrame(DataInputView in, String keyName, BiFunctionWithException<DataInputView, String, T, IOException> reader) throws IOException {
+		final int bufferSize = in.readInt();
+		byte[] buf = new byte[bufferSize];
+		in.read(buf, 0, bufferSize);
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java]: You would want to call `readFully` here to be completely resilient.
`read` only reads "up to" `#bufferSize` bytes, so there could be cases where we have some dangling bytes unread, leaving a corrupted stream.

File: flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		DataInputDeserializer frame = new DataInputDeserializer(buf, 0, bufferSize);
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java]: I don't think we need to set `startPos` / `endPos` here.
Just `new DataInputDeserializer(buf)` will do.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotData.java [hunk 6]
```diff
@@ -186,72 +185,27 @@ private static void writeDefaultKryoSerializerClasses(
...
-			}
-			catch (Throwable e) {
-				LOG.warn("Cannot deserialize a previously serialized kryo serializer for the type " + className, e);
-			}
-			kryoSerializers.put(className, javaClass, kryoSerializer);
-		}
-		return kryoSerializers;
+		return readOptionalMap(
+			in,
+			new ClassResolverByName(cl),
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotData.java]: avoid having raw types -> `new ClassResolverByName<?>(cl)`

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotTest.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	private static TypeSerializerSnapshot<Animal> kryoSnapshotWithMissingClass() throws IOException {
+		DataInputView in = new DataInputDeserializer(unLoadableSnapshotBytes());
+
+		return TypeSerializerSnapshot.readVersionedSnapshot(
+			in,
+			KryoSerializerSnapshotTest.class.getClassLoader());
+	}
+
+	/**
+	 * This method returns the bytes of a serialized {@link KryoSerializerSnapshot}, that contains a kryo registration
```
Comment (tzulitai) [inline flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerSnapshotTest.java]: all occurrences of "kryo" should have capital K

**Comment by tzulitai** [comment]
@flinkbot approve quality

**Pull request closed without merging**